src/server/scripts/Northrend/UtgardeKeep/UtgardeKeep/boss_keleseth.cpp

```diff
@@ -104,107 +98,57 @@ struct npc_frost_tomb : public NullCreatureAI
         if (victim->IsPlayer())
-        {
             Talk(SAY_KILL);
-        }
     }
 
     void JustDied(Unit* /*killer*/) override
     {
+        _JustDied();
         Talk(SAY_DEATH);
-        if (pInstance)
-            pInstance->SetData(DATA_KELESETH, DONE);
     }
 
     void JustEngagedWith(Unit* /*who*/) override
     {
-        events.Reset();
-        events.RescheduleEvent(EVENT_SPELL_SHADOWBOLT, 0ms);
-        events.RescheduleEvent(EVENT_FROST_TOMB, 28s);
-        events.RescheduleEvent(EVENT_SUMMON_SKELETONS, 4s);
-
+        _JustEngagedWith();
         Talk(SAY_START_COMBAT);
-        DoZoneInCombat();
-
-        if (pInstance)
-            pInstance->SetData(DATA_KELESETH, IN_PROGRESS);
-    }
-
-    void AttackStart(Unit* who) override
-    {
-        if (!who)
-            return;
-
-        UnitAI::AttackStartCaster(who, 12.0f);
-    }
 
-    void UpdateAI(uint32 diff) override
-    {
-        if (!UpdateVictim())
-            return;
-
-        events.Update(diff);
+        ScheduleTimedEvent(1s, [&] {
+            DoCastVictim(SPELL_SHADOWBOLT);
+        }, 4s, 5s);
 
-        if (me->HasUnitState(UNIT_STATE_CASTING))
-            return;
+        ScheduleTimedEvent(28s, [&] {
+            if (Unit* target = SelectTarget(SelectTargetMethod::Random, 0, 0.0f, true, true, -SPELL_FROST_TOMB_AURA))
+            {
+                Talk(SAY_FROST_TOMB_EMOTE, target);
+                Talk(SAY_FROST_TOMB);
+                DoCast(target, SPELL_FROST_TOMB);
+            }
+        }, 15s);
 
-        switch (events.ExecuteEvent())
-        {
-        case 0:
-            break;
-        case EVENT_SPELL_SHADOWBOLT:
-            me->CastSpell(me->GetVictim(), SPELL_SHADOWBOLT, false);
-            events.Repeat(4s, 5s);
-            break;
-        case EVENT_FROST_TOMB:
-            if (Unit* target = SelectTarget(SelectTargetMethod::Random, 0, 0.0f, true, true))
-                if (!target->HasAura(SPELL_FROST_TOMB_AURA))
-                {
-                    Talk(SAY_FROST_TOMB_EMOTE, target);
-                    Talk(SAY_FROST_TOMB);
-                    me->CastSpell(target, SPELL_FROST_TOMB, false);
-                    events.Repeat(15s);
-                    break;
-                }
-            events.Repeat(1s);
-            break;
-        case EVENT_SUMMON_SKELETONS:
+        me->m_Events.AddEventAtOffset([this]() {
             Talk(SAY_SUMMON_SKELETONS);
             for (uint8 i = 0; i < 5; ++i)
             {
                 float dist = rand_norm() * 4 + 3.0f;
                 float angle = rand_norm() * 2 * M_PI;
-                if (Creature* c = me->SummonCreature(NPC_SKELETON, 156.2f + cos(angle) * dist, 259.1f + std::sin(angle) * dist, 42.9f, 0, TEMPSUMMON_CORPSE_TIMED_DESPAWN, 20000))
-                    if (Unit* target = c->SelectNearestTarget(250.0f))
-                    {
-                        c->AddThreat(target, 5.0f);
-                        DoZoneInCombat(c);
-                    }
```

Comment:
Is this threat stuff not relevant anymore?

Reply:
No